### ISSUE TITLE
Better highlighting on errors to spot them faster

### ISSUE THREAD

**Issue by rgolea** [microsoft/vscode#16222]
Hi there!

I've been experiencing a problem when highlighting errors on VSCode. Whenever you have big issues you can spot them on the fly when opening the document. But if your issues are small, you need more time to spot those errors.

As you can see in the image below, these 2 lines have 2 errors and both of them have the problem of not having a semicolon at the end of the line:
![screen shot 2016-11-29 at 15 45 30](https://cloud.githubusercontent.com/assets/8843202/20714411/41aa18f2-b64c-11e6-8189-33bf5d188231.png)

The first one is easier to spot than the second (at the end of the parenthesis). The simpler solution for this would be to highlight the entire line in some way (so double highlight) just to spot easily the lines that have errors.

Atom does this the following way:
![screen shot 2016-11-29 at 15 48 21](https://cloud.githubusercontent.com/assets/8843202/20714571/afd983e4-b64c-11e6-931c-66c0ae556576.png)

Thanks and keep up the good work!

- VSCode Version: Version 1.7.2
- OS Version: 10.12.1 (macOS Sierra)

**Comment by hoovercj** [comment]
One solution to this would be to expose diagnostics for a document in the api. This would allow extension developers to add custom decorations for diagnostics

**Comment by rgolea** [comment]
@hoovercj Sorry, how can I do that? Or is this for the dev team?

**Comment by hoovercj** [comment]
@rgolea it is a suggestion for the dev team. It would also allow extension developers to come up with a solution for this issue that I opened  a year ago: https://github.com/Microsoft/vscode/issues/1688

**Comment by alexandrudima** [comment]
Errors and warnings can be seen through multiple views:
* outside an editor
  * Problems view (ctrl+shift+m)
* inside an editor
  * in the overview ruler (the vertical scrollbar)
  * in the editor via squiggles
  * in the editor via the F8 list

@bgashler1 Do you think we should do more with errors/warnings to make them stand out even more ?

**Comment by rgolea** [comment]
@alexandrudima the problems view is a good way to view all the errors, but a simple highlight should suffice to bother the user. For example, I am working with a team that isn't quite tidy with their code and the fact that it's highlighted is a good way to bother them and change it. If the errors are just in a simple palette or just underlining the errors, it is not really enough to make a team like that change the code. And I am sure all of us have someone in their team that does or did that.

Is there an update on this feature?

Thanks a lot guys!

**Comment by ypresto** [comment]
In dark theme, (green) warning shown via squiggles and overview ruler with scroll bar is looks like this:
<img width="79" alt="2016-12-29 18 34 16" src="https://cloud.githubusercontent.com/assets/400558/21540819/71aed4e6-cdf5-11e6-9a12-dcd66bbad4f0.png">

I think it is hard to see.

**Comment by Ciantic** [comment]
Here is a great StackOverflow that illustrates this: https://stackoverflow.com/questions/38057227/is-there-a-way-to-improve-linting-errors-highlight-in-visual-studio-code

Spot the errors! I literally spend several minutes my workday searching for this little squiggles, the Atom has figured this out. 

Image from the stack overflow question:
![image](https://user-images.githubusercontent.com/64731/27400354-83c9906a-56c8-11e7-8900-8519d68caf20.png)

**Comment by pr-yemibedu** [comment]
Hello,
 I don't many people dislike the help the editor gives, they just don't want to be limited that it is the only true way to convey information to the user. If the information is presented as a pub/sub list, then any tool, including existing problems view can take the information and present innovative visuals to cater to user needs. For example, being able to make it friendly for audio tools or to present statistics like code complexity and to maybe give better context like file type aware containment (functions or module name) and more levels of concern like **ignorable** or **bad style**. Thank you. Good day.

**Comment by steveostudios** [comment]
+1 I think this feature would be fantastic!
In Atom having a dot on the left is _way more helpful_. My eye is constantly looking at the beginning of a line when reading code. Rarely ever do I look over to the purposefully look at the right side of my screen.

**Comment by distantnative** [comment]
+1

**Comment by Valid** [comment]
+1. I'd prefer a gutter icon like the Atom example above or how ST handles it, but the squiggly just doesn't cut it for me.

**Comment by hoovercj** [comment]
I still advocate for exposing a list of all problems in a workspace via the API, both a `getDiagnostics` function and a `onDiagnosticsChanged` event.

**Comment by simonvomeyser** [comment]
+1 Would also love a gutter icon to spot errors more quickly!

**Comment by bardware** [comment]
Eclipse has these markers on the left hand side. One distinct icon for information, one for warning and one for error

**Comment by walt-er** [comment]
+1 This would be a massive help, please add! Both the option to change the highlight/underline style and, more importantly, gutter icons!

**Comment by sgmonda** [comment]
+1

**Comment by GriffinSauce** [comment]
I'd like to add an accessibility perspective: I am colourblind ([red-green](https://en.wikipedia.org/wiki/Color_blindness#Red%E2%80%93green_color_blindness), so I see colours but red and green do not pop out like you would expect) and the squiggly line is so thin that I can't make it stand out unless I make it white for max contrast. This is especially true for single character syntax errors.

These solutions would definitely work:
- the gutter dot
- being able to customize the thickness or style of the line (I'd love to make it a solid 3px underline)
- different marking options, eg. in Sublime there is the option for an outline, that is also much more discoverable

Just speaking from my own perspective, but solutions for colourblind also improve the situation for the visually impaired (or people with regular aging eyesight ;) )

**Comment by ssrballin** [comment]
+1

**Comment by balabukha** [comment]
It'd be great if you add some additional dots into code. +1

**Comment by brettshumaker** [comment]
It'd be great to be able to style the errors generated by specific extensions - i.e. making a `phpcs` extension error look different than one generated by the `PHP Intelephense` extension.

My specific use case is when dealing with legacy code (or code where I don't control the coding standards) that triggers loads of `phpcs` errors and warnings, it's nearly impossible to use the `Problems` tab. Of course I can filter for the extension I really want to see problems for, but there should be a way for users to visually distinguish problems by extension, that data is obviously available.

_Note: I think this is relevant to this issue, but I'd be happy to create my own for this feature request._

**Comment by pr-yemibedu** [comment]
Hello,
 @brettshumaker that is a helpful idea. 
  -  What would be the limit of how many different styles (like colors) would you want to see?
  -  What would be the priority for marking an extension with a  particular style (like color gradient)?
  -  What would you expect to see if two extensions see the same problem (overlap styles maybe)?

It seems interesting to have a useful version of this that would also help to know if an extension is even working or if it is being supplanted or usurped. Thank you. Good day.

**Comment by Marcelo-Petrucelli** [comment]
+1

Visualization is a key feature when dealing with errors and large codes. ^_^

**Comment by DanTup** [comment]
(FWIW, diagnostics are exposed via an API now, so it may be possible to build this as an extension?)

**Comment by chrisgreeffpp** [comment]
+1

**Comment by greevz** [comment]
+1

**Comment by werkzeugh** [comment]
+1

**Comment by usernamehw** [comment]
https://marketplace.visualstudio.com/items?itemName=PhilHindle.errorlens

![errorlens](https://user-images.githubusercontent.com/9638156/52683297-6e402780-2f53-11e9-9eba-f7225dfcb8a3.png)

https://marketplace.visualstudio.com/items?itemName=JoeBerria.statusbarerror

![statusbarerror](https://user-images.githubusercontent.com/9638156/52683560-400f1780-2f54-11e9-86be-a43dd980efcb.png)

https://marketplace.visualstudio.com/items?itemName=IgorSbitnev.error-gutters

![errorgutter](https://user-images.githubusercontent.com/9638156/52683648-81072c00-2f54-11e9-944e-a803ddc12526.png)

**Comment by isidorn** [comment]
For now are happy with how the `usernamehw.errorlens` does this. So we recommend users download this extension and try it out.
We will probably not change this in vscode core yet.

**Comment by isidorn** [comment]
Closing this as we will not make changes in vscode core, and extensions can tackle this properly on their end.

**Comment by luckylooke** [comment]
Sometimes it is very hard to spot error in Monokai theme and typescript. Imho should be better highlighted by default, not by extensions.
![Screenshot at Aug 20 06-57-25](https://user-images.githubusercontent.com/3660030/90718757-0ccce780-e2b3-11ea-9184-91c0050b1f5c.png)
